probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -0,0 +1,815 @@
+        ap: &FullyQualifiedApAddress,
+        addr: u8,
+    ) -> Result<u32, ArmError> {
+        let (index, apsel) = ap_to_bmp(ap)?;
+
+        let command = match self.probe.remote_protocol {
+            ProtocolVersion::V0 => {
+                return Err(ArmError::Probe(
+                    DebugProbeError::CommandNotSupportedByProbe {
+                        command_name: "adiv5 raw ap read",
+                    },
+                ));
+            }
+            ProtocolVersion::V0P => RemoteCommand::ReadApV0P { apsel, addr },
+            ProtocolVersion::V1 | ProtocolVersion::V2 => {
+                RemoteCommand::ReadApV1 { index, apsel, addr }
+            }
+            ProtocolVersion::V3 | ProtocolVersion::V4 => {
+                RemoteCommand::ReadApV3 { index, apsel, addr }
+            }
+        };
+        let result = u32::from_be(
+            TryInto::<u32>::try_into(
+                self.probe
+                    .command(command)
+                    .map_err(|e| ArmError::Probe(e.into()))?
+                    .0,
+            )
+            .unwrap(),
+        );
+        Ok(result)
+    }
+
+    fn write_raw_ap_register(
+        &mut self,
+        ap: &FullyQualifiedApAddress,
+        addr: u8,
+        value: u32,
+    ) -> Result<(), ArmError> {
+        let (index, apsel) = ap_to_bmp(ap)?;
+        let command = match self.probe.remote_protocol {
+            ProtocolVersion::V0 => {
+                return Err(ArmError::Probe(
+                    DebugProbeError::CommandNotSupportedByProbe {
+                        command_name: "adiv5 raw ap write",
+                    },
+                ));
+            }
+            ProtocolVersion::V0P => RemoteCommand::WriteApV0P { apsel, addr, value },
+            ProtocolVersion::V1 | ProtocolVersion::V2 => RemoteCommand::WriteApV1 {
+                index,
+                apsel,
+                addr,
+                value,
+            },
+            ProtocolVersion::V3 | ProtocolVersion::V4 => RemoteCommand::WriteApV3 {
+                index,
+                apsel,
+                addr,
+                value,
+            },
+        };
+
+        let result = self
+            .probe
+            .command(command)
+            .map_err(|e| ArmError::Probe(e.into()))?
+            .0;
+        if result == 0 {
+            Ok(())
+        } else {
+            Err(ArmError::Probe(DebugProbeError::Other(format!(
+                "probe returned unexpected result: {}",
+                result
+            ))))
+        }
+    }
+}
+
+impl ArmMemoryInterface for BlackMagicProbeMemoryInterface<'_> {
+    fn ap(&mut self) -> &mut MemoryAp {
+        &mut self.current_ap
+    }
+
+    fn base_address(&mut self) -> Result<u64, ArmError> {
+        self.current_ap.base_address(self.probe)
+    }
+
+    fn get_arm_communication_interface(
+        &mut self,
+    ) -> Result<
+        &mut crate::architecture::arm::ArmCommunicationInterface<
```

Comment:
It would be amazing to shorten those types by having an import (I think they are used quite a bit) :)